pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://tanzanite.cluster.engflow.com/invocations/default/631598e8-fc9d-4dab-be11-33e3855ea892?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on release-25.1.1-rc @ [9041a5d800f253e1fe6a87516a20d63c68c42519](https://github.com/cockroachdb/cockroach/commits/9041a5d800f253e1fe6a87516a20d63c68c42519):


```
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:513
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:610
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:120
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	Wraps: (4) ***UNEXPECTED ERROR; Failed to generate a random operation
        	            	Wraps: (5) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).randOp
        	            	  | 	pkg/workload/schemachange/operation_generator.go:218
        	            	  | [...repeated from below...]
        	            	Wraps: (6) failed generating operation: validate
        	            	Wraps: (7) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).validate
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3314
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).randOp
        	            	  | 	pkg/workload/schemachange/operation_generator.go:205
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:500
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:610
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:120
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (8) Validation FAIL:
        	            	  | id 162, db schemachange, schema schema_w0_66, name seq_w1_289: relation "seq_w1_289" (162): parent schema "schema_w0_66" (126) is dropped
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *withstack.withStack (6) *errutil.withPrefix (7) *withstack.withStack (8) *errutil.leafError
        	Test:       	TestWorkload
Schema Workload Stats
Total Schema Statements Executed = 153
Total Schema Statements Succeeded = 153
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 100
Total Transactions Rolled Back = 91
Total Transactions Executed = 191
    schema_change_external_test.go:95: error executing query="BACKUP DATABASE schemachange INTO 'nodelocal://1/backup'" args=[]: pq: failed to resolve targets specified in the BACKUP stmt: relation "seq_w1_289" (162): referenced schema ID 126: referenced descriptor not found
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload981984899
--- FAIL: TestWorkload (20.79s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #142100 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot T-sql-foundations branch-master release-blocker]
- #141813 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot T-sql-foundations branch-release-23.2]
- #141696 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2.21-rc]
- #141152 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [column &#34;crdb_internal_column_3_name_placeholder&#34; does not exist] [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1.0-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48174

--- a/pkg/sql/schemachanger/scbuild/builder_state.go
+++ b/pkg/sql/schemachanger/scbuild/builder_state.go
@@ -75,7 +75,7 @@ func (b *builderState) Ensure(e scpb.Element, target scpb.TargetStatus, meta scp
 			return
 		}
 		// Set a target for the element but check for concurrent schema changes.
-		_ = b.checkForConcurrentSchemaChanges(e)
+		_ = b.checkForConcurrentSchemaChanges(e, target)
 		b.addNewElementState(elementState{
 			element:  e,
 			initial:  scpb.Status_ABSENT,
@@ -94,7 +94,7 @@ func (b *builderState) Ensure(e scpb.Element, target scpb.TargetStatus, meta scp
 	// Check that there are no concurrent schema changes on the descriptors
 	// referenced by this element. Re-assign dst because of potential
 	// re-allocations.
-	dst = b.checkForConcurrentSchemaChanges(e)
+	dst = b.checkForConcurrentSchemaChanges(e, target)
 
 	// We were about to overwrite an element's target and metadata. Assert one
 	// disallowed case: reviving a "ghost" element, that is, add an element that
@@ -181,15 +181,29 @@ func (b *builderState) Ensure(e scpb.Element, target scpb.TargetStatus, meta scp
 	panic(errors.AssertionFailedf("unsupported incumbent target %s", oldTarget.Status()))
 }
 
-func (b *builderState) checkForConcurrentSchemaChanges(e scpb.Element) *elementState {
+func (b *builderState) checkForConcurrentSchemaChanges(
+	e scpb.Element, targetStatus scpb.TargetStatus,
+) *elementState {
 	b.ensureDescriptors(e)
 	// Check that there are no descriptors which are undergoing a concurrent
 	// schema change which might interfere with this one.
-	screl.AllTargetDescIDs(e).ForEach(func(id descpb.ID) {
+	checkID := func(id descpb.ID) {
 		if c := b.descCache[id]; c != nil && c.desc != nil && c.desc.HasConcurrentSchemaChanges() {
 			panic(scerrors.ConcurrentSchemaChangeError(c.desc))
 		}
-	})
+	}
+	screl.AllTargetDescIDs(e).ForEach(checkID)
+	// For new namespace elements we need to also check their parents
+	// are not in middle of a schema change. Otherwise, it's possible to
+	// add an new namespace entry inside a dropped SCHEMA or DATABASE.
+	if namespace, ok := (e).(*scpb.Namespace); ok && targetStatus == scpb.ToPublic {
+		if namespace.DatabaseID != descpb.InvalidID {
+			checkID(namespace.DatabaseID)
+		}
+		if namespace.SchemaID != descpb.InvalidID {
+			checkID(namespace.SchemaID)
+		}
+	}
 	// We may have mutated the builder state for this element.
 	// Specifically, the output slice might have grown and have been realloc'ed.
 	return b.getExistingElementState(e)

--- a/pkg/sql/schemachanger/schemachanger_test.go
+++ b/pkg/sql/schemachanger/schemachanger_test.go
@@ -1105,3 +1105,44 @@ CREATE TABLE t2(n int);
 		})
 	}
 }
+
+// TestPreventCreateDropConcurrently confirms that objects cannot be
+// created under a schema that is being dropped.
+func TestPreventCreateDropConcurrently(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
+		// This would work with secondary tenants as well, but the span config
+		// limited logic can hit transaction retries on the span_count table.
+		DefaultTestTenant: base.TestIsForStuffThatShouldWorkWithSecondaryTenantsButDoesntYet(138733),
+	})
+	defer s.Stopper().Stop(ctx)
+
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+
+	runner.Exec(t, `
+CREATE SCHEMA other_schema;
+CREATE SCHEMA complex_drop_schema;
+CREATE TABLE complex_drop_schema.t1(n int UNIQUE);
+CREATE TABLE other_schema.t1(n int REFERENCES complex_drop_schema.t1(n));
+`)
+
+	runner.Exec(t, `SET CLUSTER SETTING jobs.debug.pausepoints = 'newschemachanger.before.exec';`)
+	runner.ExpectErr(t, " \\d+ was paused before it completed with reason: pause point \"newschemachanger.before.exec\" hit",
+		"DROP SCHEMA complex_drop_schema CASCADE;")
+
+	grp := ctxgroup.WithContext(ctx)
+	grp.GoCtx(func(ctx context.Context) error {
+		_, err := sqlDB.Exec("CREATE SEQUENCE  complex_drop_schema.sc1")
+		return err
+	})
+
+	runner.Exec(t, `SET CLUSTER SETTING jobs.debug.pausepoints = ''`)
+	runner.Exec(t,
+		`RESUME JOB (SELECT job_id FROM crdb_internal.jobs WHERE description LIKE 'DROP SCHEMA%' AND status='paused' FETCH FIRST 1 ROWS ONLY);`)
+	require.Error(t,
+		grp.Wait(),
+		`cannot create "complex_drop_schema.sc1" because the target database or schema does not exist`)
+}

